### PR TITLE
Issue #913: Log NLog version to internal log.

### PR DESCRIPTION
**Pull request by bhaeussermann**

### CODE REVIEW

File: src/NLog/Common/InternalLoggerHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    /// </summary>
+    public static class InternalLoggerHelpers
+    {
+        /// <summary>
+        /// Logs the assembly version and file version of the given Assembly.
+        /// </summary>
+        /// <param name="assembly">The assembly to log.</param>
+        public static void LogAssemblyVersion(Assembly assembly)
+        {
+            InternalLogger.Info("{0} - Assembly version: {1}; File version: {2}.",
```
Comment (304NotModified) [inline src/NLog/Common/InternalLoggerHelpers.cs]: 1. can you add a try-catch around this? To be sure. 
2. we also like the AssemblyInformationalVersion  (System.Diagnostics.FileVersionInfo.ProductVersion) 

thanks!

File: src/NLog/Config/ConfigurationItemFactory.cs [hunk 2]
```diff
@@ -104,7 +98,16 @@ public ConfigurationItemFactory(params Assembly[] assemblies)
         /// <summary>
         /// Gets or sets default singleton instance of <see cref="ConfigurationItemFactory"/>.
         /// </summary>
-        public static ConfigurationItemFactory Default { get; set; }
+        public static ConfigurationItemFactory Default
+        {
+            get
+            {
+                if (defaultInstance == null)
```
Comment (304NotModified) [inline src/NLog/Config/ConfigurationItemFactory.cs]: why this change?

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 3]
```diff
@@ -65,10 +65,14 @@ public class XmlLoggingConfiguration : LoggingConfiguration
     {
         #region private fields
 
-        private readonly ConfigurationItemFactory configurationItemFactory = ConfigurationItemFactory.Default;
         private readonly Dictionary<string, bool> visitedFile = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
 
         private string originalFileName;
+        
+        private ConfigurationItemFactory ConfigurationItemFactory
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: why this change?

File: src/NLog/Common/InternalLoggerHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+
+namespace NLog.Common
+{
+    /// <summary>
+    /// Helper methods for internal logger.
+    /// </summary>
+    public static class InternalLoggerHelpers
```
Comment (304NotModified) [inline src/NLog/Common/InternalLoggerHelpers.cs]: I'm curious why this isn't added to the `InternalLogger` class?

**Comment by 304NotModified** [comment]
Thanks for this PR :+1: 

I reviewed your code. I looks good, but I saw some unexpected changes. See the code notes :)

**Comment by 304NotModified** [comment]
I think `FileVersionInfo` isn't in Silverlight (see AppVeyor message). 

Please add for silverlight the log, but not the fileversion. You can use `#if SILVERLIGHT` , `#else`, and `#endif`

See https://github.com/NLog/NLog/blob/master/CONTRIBUTING.md#multiple-net-versions

File: src/NLog/Config/ConfigurationItemFactory.cs (hunk 2 above)
Comment (bhaeussermann) [inline src/NLog/Config/ConfigurationItemFactory.cs]: The first commit of the pull request ("Fixed Auto-loading logs did not render because extension DLL's were loaded before InternalLogger config was read") contains all the changes that were needed to fix the auto-loading logs that did not appear.

This part, in particular, implements lazy instantiation on the ConfigurationItemFactory, because it was created too early. Now the ConfigurationItemFactory is created only after the internal logger had been configured, so that the auto-loading logs actually appear in the internal log file.

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 3 above)
Comment (bhaeussermann) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: This is connected to the other change you queried. It facilitates the lazy instantiation of the `ConfigurationItemFactory`.

File: src/NLog/Common/InternalLoggerHelpers.cs (hunk 4 above)
Comment (bhaeussermann) [inline src/NLog/Common/InternalLoggerHelpers.cs]: The logic seemed too specific to add to `InternalLogger`.  
Would you think this to fit better in `InternalLogger`?

Comment (304NotModified) [inline src/NLog/Common/InternalLoggerHelpers.cs]: I prefer combining with InternalLogger with these reasons:

- It's highly correlated to the InternalLogger (25%)
- The method is not used somewhere else
- The method can be created `private` after combine, which has preference (= less maintenance)
- The InternalLogger has now only one method.

File: src/NLog/Config/ConfigurationItemFactory.cs (hunk 2 above)
Comment (304NotModified) [inline src/NLog/Config/ConfigurationItemFactory.cs]: OK thanks. Can you add this (summary is OK), to the `<remarks>` ?

**Comment by bhaeussermann** [comment]
We now write `assembly.FullName` to the log, which includes the assembly version. The log now looks like this:
`2015-10-05 20:24:08.3910 Info NLog.Web, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null. File version: 2.0.14. Product version: 2.0.14.`

**Comment by 304NotModified** [comment]
Good idea!

**Comment by bhaeussermann** [comment]
The pull request is ready to my knowledge. Do you notice anything that's outstanding?

**Comment by 304NotModified** [comment]
I looks good. I will check the PR in detal this week

**Comment by 304NotModified** [comment]
Thanks for this nice PR! :+1:
